python/cuda_parallel/cuda/parallel/experimental/struct.py

```diff
@@ -0,0 +1,133 @@
+from dataclasses import dataclass
+from dataclasses import fields as dataclass_fields
+from typing import Type
+
+import numba
+import numpy as np
+from numba.core import cgutils
+from numba.core.extending import (
+    make_attribute_wrapper,
+    models,
+    register_model,
+    typeof_impl,
+)
+from numba.core.typing import signature as nb_signature
+from numba.core.typing.templates import AttributeTemplate, ConcreteTemplate
+from numba.cuda.cudadecl import registry as cuda_registry
+from numba.extending import lower_builtin
+
+from .typing import GpuStruct
+
+
+def gpu_struct(this: type) -> Type[GpuStruct]:
+    """
+    Defines the given class as being a GpuStruct.
+
+    A GpuStruct represents a value composed of one or more other
+    values, and is defined as a class with annotated fields (similar
+    to a dataclass). The type of each field must be a subclass of
+    `np.number`, like `np.int32` or `np.float64`.
+
+    Arrays of GPUStruct objects can be used as inputs to algorithms.
+
+    Example:
+        The code snippet below shows how to use `gpu_struct` to define
+        a `Pixel` type (composed of `r`, `g` and `b` values),  and perform
+        a reduction on an array of `Pixel` objects to identify the one
+        with the largest `g` component:
+
+        .. literalinclude:: ../../python/cuda_parallel/tests/test_reduce_api.py
+            :language: python
+            :dedent:
+            :start-after: example-begin reduce-struct
+            :end-before: example-end reduce-struct
+    """
+    # Implementation-wise, @gpu_struct creates and registers a
+    # corresponding numba type to the given type, so that it can be
+    # used within device functions (e.g., unary and binary operations)
+    # The numba typing/lowering code is largely based on the example
+    # in https://github.com/gmarkall/numba-accelerated-udfs.
+
+    anns = getattr(this, "__annotations__", {})
+
+    # Set a .dtype attribute on the class that returns the
+    # corresponding numpy structure dtype. This makes it convenient to
+    # create CuPy/NumPy arrays of this type.
+    setattr(this, "dtype", np.dtype(list(anns.items())))
+
+    # Define __post_init__ to create a ctypes object from the fields,
+    # and keep a reference to it in the `._data` attribute. This
+    # ctypes object is what is ultimately passed to the underlying C
+    # library and we need to make sure we keep it alive.
```

Comment:
not sure I understand the reason that ctypes is needed here, couldn't we just allocate a size-1 numpy array?
```python
self._data = np.empty(1, dtype=this.dtype)
```
then in `to_cccl_value()` you don't need the if-else branch.

Reply:
(I think it should be also possible to populate `self._data` in this case.)

Reply:
Something like this? https://github.com/NVIDIA/cccl/pull/3218/commits/1654415e302b4673e259d60206fba730d760ffdb